src/main/java/org/xstefank/model/PersistentList.java

```diff
@@ -30,32 +34,136 @@ public PersistentList(String dirName, String fileName) {
         this(new File(dirName, fileName));
     }
 
-    public void addUser(String username) {
-        if (list.contains(username)) {
-            return;
+    @Override
+    public String set(int index, String element) {
+        String oldValue = super.set(index, element);
+        rewriteFileContent(index);
+        return oldValue;
+    }
+
+    @Override
+    public boolean add(String s) {
+        if (contains(s)) {
+            return false;
         }
-        list.add(username);
         try (FileWriter fw = new FileWriter(file, true)) {
-            fw.write(username + "\n");
+            fw.write(s + "\n");
+            return super.add(s);
```

Comment:
you should stay consistent with ordering of operations. Everywhere else you first modify in-memory list and then write file except here